sys/include/usb/dfu.h

```diff
@@ -0,0 +1,176 @@
+
+/**
+ * @name    Default USB detach timeout for DFU descriptor
+ * @{
+ */
+#ifndef USB_DFU_DETACH_TIMEOUT_MS
+#define USB_DFU_DETACH_TIMEOUT_MS 255
+#endif
+/** @} */
+
+
+/**
+ * @name USB DFU interface attributes
+ * @{
+ */
+#define USB_DFU_CAN_DOWNLOAD            0x01    /**< DFU Download attribute */
+#define USB_DFU_CAN_UPLOAD              0x02    /**< DFU Upload attribute */
+#define USB_DFU_MANIFEST_TOLERANT       0x04    /**< DFU Manifest tolerant attribute */
+#define USB_DFU_WILL_DETACH             0x08    /**< DFU Detach capabability attribute */
+/** @} */
+/**
+ * @name USB DFU interface type
+ * @{
+ */
+#define USB_DFU_INTERFACE               0xFE /** Application Specific Interface */
+/** @} */
+/**
+ * @name USB DFU subclass types
+ * @anchor usb_dfu_subtype
+ * @{
+ */
+#define USB_DFU_SUBCLASS_DFU            0x01 /**< DFU subclass */
+/** @} */
+
+/**
+ * @name USB DFU protocol types
+ * @{
+ */
+#define USB_DFU_PROTOCOL_RUNTIME_MODE   0x01 /**< Runtime mode */
+#define USB_DFU_PROTOCOL_DFU_MODE       0x02 /**< DFU mode */
+/** @} */
+
+/**
+ * @name USB DFU setup request
+ * @{
+ */
+#define DFU_DETACH      0x00        /**< DFU Detach request */
+#define DFU_DOWNLOAD    0x01        /**< DFU Download request */
+#define DFU_UPLOAD      0x02        /**< DFU Upload request */
+#define DFU_GET_STATUS  0x03        /**< DFU Get Status request */
+#define DFU_CLR_STATUS  0x04        /**< DFU Clear Status request */
+#define DFU_GET_STATE   0x05        /**< DFU Get State request */
+#define DFU_ABORT       0x06        /**< DFU Abort request */
+/** @} */
+
+/**
+ * @brief USBUS DFU internal state
+ */
+typedef enum {
+    USBUS_DFU_STATE_APP_IDLE,               /**< DFU application idle */
+    USBUS_DFU_STATE_APP_DETACH,             /**< DFU application detach (reboot to DFU mode) */
+    USBUS_DFU_STATE_DFU_IDLE,               /**< DFU runtime mode idle */
+    USBUS_DFU_STATE_DFU_DL_SYNC,            /**< DFU download synchronization */
+    USBUS_DFU_STATE_DFU_DL_BUSY,            /**< DFU download busy */
+    USBUS_DFU_STATE_DFU_DL_IDLE,            /**< DFU download idle */
+    USBUS_DFU_STATE_DFU_MANIFEST_SYNC,      /**< DFU manifest synchronization */
+    USBUS_DFU_STATE_DFU_MANIFEST,           /**< DFU manifest mode */
+    USBUS_DFU_STATE_DFU_MANIFEST_WAIT_RST,  /**< DFU manifest wait for CPU reset */
+    USBUS_DFU_STATE_DFU_UP_IDLE,            /**< DFU upload idle */
+    USBUS_DFU_STATE_DFU_ERROR               /**< DFU internal error */
+} usbus_dfu_state_t;
+
+/**
+ * @brief USB DFU interface descriptor
+ */
+typedef struct __attribute__((packed)) {
+    uint8_t length;              /**< Descriptor length */
+    uint8_t type;                /**< Descriptor type */
+    uint8_t attribute;           /**< Descriptor attributes flags */
+    uint16_t detach_timeout;     /**< Descriptor detach timeout (ms) */
+    uint16_t xfer_size;          /**< Descriptor transfer size */
```

Comment:
Bit of a nitpick, but this matches the terminology used by the spec:

```suggestion
    uint16_t xfer_size;          /**< Descriptor transaction size */
```